{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9e5e625",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "x is a <class 'int'> and is set to: 10\n",
      "p1 is a <class 'float'> and is set to: 3.1415\n",
      "gene is a <class 'str'> and is set to: MAPT\n",
      "is_neuron is a <class 'bool'> and is set to: True\n",
      "\n",
      "Gene list should be: MAPT, SNCA, APP, C9orf72, GRN as called by 1,2,0,4,3 positions... and, it is:\n",
      "MAPT in index position 1\n",
      "SNCA in index position 2\n",
      "APP in index position 0\n",
      "C9orf72 in index position 4\n",
      "GRN in index position 3\n",
      "\n",
      "GRN gene added as 9.7. It is: 9.7\n",
      "All dictionary pairs: dict_items([('APP', 12.3), ('MAPT', 8.5), ('SNCA', 15.2), ('GRN', 9.7)])\n",
      "\n",
      "MAPT shows 1.8-fold upregulation\n",
      "\n",
      "gene list with C9orf72 appended previously:\n",
      "APP in index 0\n",
      "MAPT in index 1\n",
      "SNCA in index 2\n",
      "GRN in index 3\n",
      "C9orf72 in index 4\n",
      "\n",
      "gene list (keys) and values in expr dictionary:\n",
      "APP: 12.3\n",
      "MAPT: 8.5\n",
      "SNCA: 15.2\n",
      "GRN: 9.7\n",
      "\n",
      "Final gene expression summary (in transcripts per million):\n",
      "\n",
      "1. APP      => 12.30 TPM\n",
      "2. MAPT     =>  8.50 TPM\n",
      "3. SNCA     => 15.20 TPM\n",
      "4. GRN      =>  9.70 TPM\n",
      "5. C9orf72  => 11.40 TPM\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# week1_day1_basics\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "# 1) No variable declaration needed--python fills in types automatically\n",
    "x = 10           # int\n",
    "pi = 3.1415      # float\n",
    "gene = \"MAPT\"    # string\n",
    "is_neuron = True # boolean\n",
    "\n",
    "# check variable types\n",
    "print (f\"x is a {type(x)} and is set to: {x}\")\n",
    "print (f\"p1 is a {type(pi)} and is set to: {pi}\")\n",
    "print (f\"gene is a {type(gene)} and is set to: {gene}\")\n",
    "print (f\"is_neuron is a {type(is_neuron)} and is set to: {is_neuron}\")\n",
    "print()\n",
    "\n",
    "# 2) Lists are changeable and Tuples are not changable\n",
    "genes = [\"APP\", \"MAPT\", \"SNCA\", \"GRN\"]\n",
    "genes.append(\"C9orf72\")\n",
    "print (f\"Gene list should be: MAPT, SNCA, APP, C9orf72, GRN as called by 1,2,0,4,3 positions... and, it is:\")\n",
    "indicies = [1,2,0,4,3]\n",
    "for i in indicies:\n",
    "    print (f\"{genes[i]} in index position {i}\")   # MAPT, SNCA, APP, C9orf72, GRN\n",
    "print ()\n",
    "\n",
    "# 3) Dictionaries (key-value pairs, and can also be lists/arrays of pairs)\n",
    "expr = {\"APP\":12.3, \"MAPT\":8.5, \"SNCA\":15.2}      \n",
    "expr[\"GRN\"] = 9.7           # add new key\n",
    "print(f\"GRN gene added as 9.7. It is: {expr['GRN']}\")\n",
    "print(f\"All dictionary pairs: {expr.items()}\")\n",
    "# Test to see if dictionary entries also be called by list locations?  --no. You can loop through the list iteratively, but the index location is not numbered like a list. It is indexed by key name, not number. \n",
    "print ()\n",
    "\n",
    "# 4) Combine these elements together to call variables with for-loops rather than individually \n",
    "gene = \"MAPT\"\n",
    "fold_change = 1.8\n",
    "print(f\"{gene} shows {fold_change}-fold upregulation\")\n",
    "print ()\n",
    "\n",
    "# Show list of genes using a simple for-loop, rather than calling each individually\n",
    "print (f\"gene list with C9orf72 appended previously:\")\n",
    "for g in genes:\n",
    "    print(f\"{g} in index {genes.index(g)}\") \n",
    "print()\n",
    "\n",
    "# Dictionary lookup using a simple for-loop, rather than calling each individually\n",
    "expr = {\"APP\":12.3, \"MAPT\":8.5, \"SNCA\":15.2, \"GRN\":9.7}\n",
    "print (f\"gene list (keys) and values in expr dictionary:\")\n",
    "for k, v in expr.items():\n",
    "    print(f\"{k}: {v}\")\n",
    "#How to print the key/name of each pair? and not just the value? -- ans: call the name as expr.key and the value as expr.value, and the pair as expr.items (returns both key and value)  \n",
    "print()\n",
    "\n",
    "# Combining it all together into a formatted table using enumerate (takes any iterable (like a list, tuple, or string) and returns an iterator that produces pairs of index, value)\n",
    "# First, add C9orf72 key and value to expr dictionary\n",
    "expr[\"C9orf72\"] = 11.4\n",
    "#Then print the full output of the dictionary using the 'genes' list as they 'key'/name source (for no real reason other than to confirm it matches the dictionary)\n",
    "#Also play with some nicer formating in the output\n",
    "print(\"Final gene expression summary (in transcripts per million):\\n\")\n",
    "for i, g in enumerate(genes):\n",
    "    print(f\"{i+1}. {g:8s} => {expr[g]:5.2f} TPM\")\n",
    "print()\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
