{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c1c85daa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.3.2\n",
      "      Gene  Expression Condition\n",
      "0      APP        14.2        AD\n",
      "1     MAPT         9.8        AD\n",
      "2     SNCA        16.3        PD\n",
      "3      GRN         8.7       FTD\n",
      "4  C9orf72        10.9       FTD\n",
      "\n",
      "      Gene  Expression Condition   High\n",
      "2     SNCA        16.3        PD   True\n",
      "0      APP        14.2        AD   True\n",
      "4  C9orf72        10.9       FTD   True\n",
      "1     MAPT         9.8        AD  False\n",
      "3      GRN         8.7       FTD  False\n",
      "\n",
      "      Gene  Expression Condition   High\n",
      "0     SNCA        16.3        PD   True\n",
      "1      APP        14.2        AD   True\n",
      "2  C9orf72        10.9       FTD   True\n",
      "3     MAPT         9.8        AD  False\n",
      "4      GRN         8.7       FTD  False\n",
      "\n",
      "Saved DataFrame from .csv file:\n",
      "      Gene  Expression Condition   High\n",
      "0     SNCA        16.3        PD   True\n",
      "1      APP        14.2        AD   True\n",
      "2  C9orf72        10.9       FTD   True\n",
      "3     MAPT         9.8        AD  False\n",
      "4      GRN         8.7       FTD  False\n"
     ]
    }
   ],
   "source": [
    "#week2_day1_pandas_intro\n",
    "\n",
    "import pandas as pd\n",
    "print(pd.__version__)\n",
    "\n",
    "#create a DataFrame with 3 series and then print it out with a numbered row\n",
    "data = {\n",
    "    \"Gene\": [\"APP\", \"MAPT\", \"SNCA\", \"GRN\", \"C9orf72\"],\n",
    "    \"Expression\": [14.2, 9.8, 16.3, 8.7, 10.9],\n",
    "    \"Condition\": [\"AD\", \"AD\", \"PD\", \"FTD\", \"FTD\"]\n",
    "}\n",
    "\n",
    "df = pd.DataFrame(data)\n",
    "print(df)\n",
    "print()\n",
    "\n",
    "## some example commands for DataFrames\n",
    "#df.head()      # prints first 5 rows\n",
    "#df.tail(2)     # prints last 2 rows\n",
    "#df.info()      # prints out structure, types, and counts\n",
    "#df.describe()  # summary stats\n",
    "\n",
    "# Syntax for calling DF indexes and checks\n",
    "#df[df[\"Expression\"] > 10]\n",
    "#df[(df[\"Expression\"] > 10) & (df[\"Condition\"] == \"AD\")]\n",
    "\n",
    "#Adds a new column of just highly-expressed rows (>10)\n",
    "df[\"High\"] = df[\"Expression\"] > 10\n",
    "#print(df[\"High\"])\n",
    "\n",
    "#drop (delete) a column\n",
    "#df = df.drop(columns=[\"Condition\"])\n",
    "\n",
    "# sort or index whole DFs by column values or other\n",
    "#df.sort_values() can take the following arguments: (by=\"name\", axis=(0 is rows, 1 is columns), ascending=True, inplace=False (False returns a new sorted DF and leaves the original alone), kind='quicksort', na_position='last'(where to place blank or missing values))\n",
    "df.sort_values(by=\"Expression\", ascending=False, inplace=True)\n",
    "print(df)\n",
    "print()\n",
    "\n",
    "#df.reset_index will set the index to be 0-N based on the current sorted order\n",
    "df.reset_index(drop=True, inplace=True)\n",
    "print(df)\n",
    "print()\n",
    "\n",
    "# DF.to_csv will save the DF to a .csv file (in the same folder, in this case)\n",
    "df.to_csv(\"E:/DATA_SCIENCE/Projects/gene_expression.csv\", index=False)\n",
    "df2 = pd.read_csv(\"E:/DATA_SCIENCE/Projects/gene_expression.csv\")\n",
    "print(f\"Saved DataFrame from .csv file:\")\n",
    "print(df2)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "959627e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.3.2\n",
      "      Gene  Ct_Control  Ct_Treated\n",
      "0      APP        23.1        21.8\n",
      "1     MAPT        26.4        25.7\n",
      "2     SNCA        24.9        26.2\n",
      "3      GRN        27.5        26.0\n",
      "4  C9orf72        29.1        27.9\n",
      "\n",
      "4 new columns added: \n",
      "      Gene  Ct_Control  Ct_Treated  dCt  mean_dCt  ddCt  fold_change\n",
      "0      APP        23.1        21.8 -1.3     -0.68 -0.62     1.536875\n",
      "1     MAPT        26.4        25.7 -0.7     -0.68 -0.02     1.013959\n",
      "2     SNCA        24.9        26.2  1.3     -0.68  1.98     0.253490\n",
      "3      GRN        27.5        26.0 -1.5     -0.68 -0.82     1.765406\n",
      "4  C9orf72        29.1        27.9 -1.2     -0.68 -0.52     1.433955\n",
      "\n",
      "Check for high or low fold change: \n",
      "      Gene  Ct_Control  Ct_Treated  dCt  mean_dCt  ddCt  fold_change   High  \\\n",
      "0      APP        23.1        21.8 -1.3     -0.68 -0.62     1.536875   True   \n",
      "1     MAPT        26.4        25.7 -0.7     -0.68 -0.02     1.013959  False   \n",
      "2     SNCA        24.9        26.2  1.3     -0.68  1.98     0.253490  False   \n",
      "3      GRN        27.5        26.0 -1.5     -0.68 -0.82     1.765406   True   \n",
      "4  C9orf72        29.1        27.9 -1.2     -0.68 -0.52     1.433955   True   \n",
      "\n",
      "     Low  \n",
      "0  False  \n",
      "1  False  \n",
      "2   True  \n",
      "3  False  \n",
      "4  False  \n",
      "\n",
      "Sorted by fold change and re-indexed: \n",
      "\n",
      "Saved qPCR DataFrame from .csv file:\n",
      "      Gene  Ct_Control  Ct_Treated  dCt  mean_dCt  ddCt  fold_change   High  \\\n",
      "0      GRN        27.5        26.0 -1.5     -0.68 -0.82     1.765406   True   \n",
      "1      APP        23.1        21.8 -1.3     -0.68 -0.62     1.536875   True   \n",
      "2  C9orf72        29.1        27.9 -1.2     -0.68 -0.52     1.433955   True   \n",
      "3     MAPT        26.4        25.7 -0.7     -0.68 -0.02     1.013959  False   \n",
      "4     SNCA        24.9        26.2  1.3     -0.68  1.98     0.253490  False   \n",
      "\n",
      "     Low regulation  \n",
      "0  False         up  \n",
      "1  False         up  \n",
      "2  False         up  \n",
      "3  False         up  \n",
      "4   True       down  \n"
     ]
    }
   ],
   "source": [
    "# Day 4 Challenge — “Mock qPCR Gene Expression Analysis”\n",
    "# Goal:\n",
    "# load a small dataset of mock qPCR results, clean it, analyze differential expression, and export a summary report using Pandas.\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "print(pd.__version__)\n",
    "\n",
    "PCRdata = {\n",
    "    \"Gene\": [\"APP\", \"MAPT\", \"SNCA\", \"GRN\", \"C9orf72\"],\n",
    "    \"Ct_Control\": [23.1, 26.4, 24.9, 27.5, 29.1],\n",
    "    \"Ct_Treated\": [21.8, 25.7, 26.2, 26.0, 27.9]\n",
    "}\n",
    "PCRdf = pd.DataFrame(PCRdata)\n",
    "print(PCRdf)\n",
    "print()\n",
    "\n",
    "#Compute ΔCt and ΔΔCt\n",
    "PCRdf[\"dCt\"] = PCRdf[\"Ct_Treated\"] - PCRdf[\"Ct_Control\"]\n",
    "PCRdf[\"mean_dCt\"] = PCRdf[\"dCt\"].mean()\n",
    "PCRdf[\"ddCt\"] = PCRdf[\"dCt\"] - PCRdf[\"mean_dCt\"]\n",
    "PCRdf[\"fold_change\"] = (np.power(2, -PCRdf[\"ddCt\"]))   # np.power(base, exponent) --> exponent calculation from numpy library\n",
    "\n",
    "print(\"4 new columns added: \")\n",
    "print(PCRdf)\n",
    "print()\n",
    "\n",
    "PCRdf[\"High\"] = PCRdf[\"fold_change\"] > 1.2\n",
    "PCRdf[\"Low\"] = PCRdf[\"fold_change\"] < 0.8\n",
    "\n",
    "print(\"Check for high or low fold change: \")\n",
    "print(PCRdf)\n",
    "print()\n",
    "\n",
    "PCRdf.sort_values(by=\"fold_change\", ascending=False, inplace=True)\n",
    "PCRdf.reset_index(drop=True, inplace=True)\n",
    "PCRdf[\"fold_change\"].round(2)\n",
    "\n",
    "PCRdf[\"regulation\"] = np.where(\n",
    "    PCRdf[\"fold_change\"] > 1, \"up\",\n",
    "    np.where(PCRdf[\"fold_change\"] < 1, \"down\", \"no change\")\n",
    ")\n",
    "\n",
    "print(\"Sorted by fold change and re-indexed: \")\n",
    "PCRdf.style.background_gradient(cmap=\"coolwarm\")\n",
    "print()\n",
    "\n",
    "PCRdf.to_csv(\"E:/DATA_SCIENCE/Projects/qPCR_gene_expression.csv\", index=False)\n",
    "PCRdf_filtered = pd.read_csv(\"E:/DATA_SCIENCE/Projects/qPCR_gene_expression.csv\")\n",
    "print(f\"Saved qPCR DataFrame from .csv file:\")\n",
    "print(PCRdf_filtered)\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
