{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c1c85daa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.3.2\n",
      "      Gene  Expression Condition\n",
      "0      APP        14.2        AD\n",
      "1     MAPT         9.8        AD\n",
      "2     SNCA        16.3        PD\n",
      "3      GRN         8.7       FTD\n",
      "4  C9orf72        10.9       FTD\n",
      "\n",
      "      Gene  Expression Condition   High\n",
      "2     SNCA        16.3        PD   True\n",
      "0      APP        14.2        AD   True\n",
      "4  C9orf72        10.9       FTD   True\n",
      "1     MAPT         9.8        AD  False\n",
      "3      GRN         8.7       FTD  False\n",
      "\n",
      "      Gene  Expression Condition   High\n",
      "0     SNCA        16.3        PD   True\n",
      "1      APP        14.2        AD   True\n",
      "2  C9orf72        10.9       FTD   True\n",
      "3     MAPT         9.8        AD  False\n",
      "4      GRN         8.7       FTD  False\n",
      "\n",
      "Saved DataFrame from .csv file:\n",
      "      Gene  Expression Condition   High\n",
      "0     SNCA        16.3        PD   True\n",
      "1      APP        14.2        AD   True\n",
      "2  C9orf72        10.9       FTD   True\n",
      "3     MAPT         9.8        AD  False\n",
      "4      GRN         8.7       FTD  False\n"
     ]
    }
   ],
   "source": [
    "#week2_day1_pandas_intro\n",
    "\n",
    "import pandas as pd\n",
    "print(pd.__version__)\n",
    "\n",
    "#create a DataFrame with 3 series and then print it out with a numbered row\n",
    "data = {\n",
    "    \"Gene\": [\"APP\", \"MAPT\", \"SNCA\", \"GRN\", \"C9orf72\"],\n",
    "    \"Expression\": [14.2, 9.8, 16.3, 8.7, 10.9],\n",
    "    \"Condition\": [\"AD\", \"AD\", \"PD\", \"FTD\", \"FTD\"]\n",
    "}\n",
    "\n",
    "df = pd.DataFrame(data)\n",
    "print(df)\n",
    "print()\n",
    "\n",
    "## some example commands for DataFrames\n",
    "#df.head()      # prints first 5 rows\n",
    "#df.tail(2)     # prints last 2 rows\n",
    "#df.info()      # prints out structure, types, and counts\n",
    "#df.describe()  # summary stats\n",
    "\n",
    "# Syntax for calling DF indexes and checks\n",
    "#df[df[\"Expression\"] > 10]\n",
    "#df[(df[\"Expression\"] > 10) & (df[\"Condition\"] == \"AD\")]\n",
    "\n",
    "#Adds a new column of just highly-expressed rows (>10)\n",
    "df[\"High\"] = df[\"Expression\"] > 10\n",
    "#print(df[\"High\"])\n",
    "\n",
    "#drop (delete) a column\n",
    "#df = df.drop(columns=[\"Condition\"])\n",
    "\n",
    "# sort or index whole DFs by column values or other\n",
    "#df.sort_values() can take the following arguments: (by=\"name\", axis=(0 is rows, 1 is columns), ascending=True, inplace=False (False returns a new sorted DF and leaves the original alone), kind='quicksort', na_position='last'(where to place blank or missing values))\n",
    "df.sort_values(by=\"Expression\", ascending=False, inplace=True)\n",
    "print(df)\n",
    "print()\n",
    "\n",
    "#df.reset_index will set the index to be 0-N based on the current sorted order\n",
    "df.reset_index(drop=True, inplace=True)\n",
    "print(df)\n",
    "print()\n",
    "\n",
    "# DF.to_csv will save the DF to a .csv file (in the same folder, in this case)\n",
    "df.to_csv(\"E:/DATA_SCIENCE/Projects/gene_expression.csv\", index=False)\n",
    "df2 = pd.read_csv(\"E:/DATA_SCIENCE/Projects/gene_expression.csv\")\n",
    "print(f\"Saved DataFrame from .csv file:\")\n",
    "print(df2)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "959627e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.3.2\n",
      "      Gene  Ct_Control  Ct_Treated\n",
      "0      APP        23.1        21.8\n",
      "1     MAPT        26.4        25.7\n",
      "2     SNCA        24.9        26.2\n",
      "3      GRN        27.5        26.0\n",
      "4  C9orf72        29.1        27.9\n",
      "\n",
      "4 new columns added: \n",
      "      Gene  Ct_Control  Ct_Treated  dCt  mean_dCt  ddCt  fold_change\n",
      "0      APP        23.1        21.8 -1.3     -0.68 -0.62     1.536875\n",
      "1     MAPT        26.4        25.7 -0.7     -0.68 -0.02     1.013959\n",
      "2     SNCA        24.9        26.2  1.3     -0.68  1.98     0.253490\n",
      "3      GRN        27.5        26.0 -1.5     -0.68 -0.82     1.765406\n",
      "4  C9orf72        29.1        27.9 -1.2     -0.68 -0.52     1.433955\n",
      "\n",
      "Check for high or low fold change: \n",
      "      Gene  Ct_Control  Ct_Treated  dCt  ...  ddCt  fold_change   High    Low\n",
      "0      APP        23.1        21.8 -1.3  ... -0.62     1.536875   True  False\n",
      "1     MAPT        26.4        25.7 -0.7  ... -0.02     1.013959  False  False\n",
      "2     SNCA        24.9        26.2  1.3  ...  1.98     0.253490  False   True\n",
      "3      GRN        27.5        26.0 -1.5  ... -0.82     1.765406   True  False\n",
      "4  C9orf72        29.1        27.9 -1.2  ... -0.52     1.433955   True  False\n",
      "\n",
      "[5 rows x 9 columns]\n",
      "\n",
      "Sorted by fold change and re-indexed: \n",
      "\n",
      "qPCR DataFrame recalled from saved .csv file:\n",
      "      Gene  Ct_Control  Ct_Treated  dCt  ...  fold_change   High    Low  regulation\n",
      "0      GRN        27.5        26.0 -1.5  ...     1.765406   True  False          up\n",
      "1      APP        23.1        21.8 -1.3  ...     1.536875   True  False          up\n",
      "2  C9orf72        29.1        27.9 -1.2  ...     1.433955   True  False          up\n",
      "3     MAPT        26.4        25.7 -0.7  ...     1.013959  False  False          up\n",
      "4     SNCA        24.9        26.2  1.3  ...     0.253490  False   True        down\n",
      "\n",
      "[5 rows x 10 columns]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_9758f_row0_col6 {\n",
       "  background-color: #b40426;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_9758f_row1_col6 {\n",
       "  background-color: #e36c55;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_9758f_row2_col6 {\n",
       "  background-color: #f18d6f;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_9758f_row3_col6 {\n",
       "  background-color: #dddcdc;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_9758f_row4_col6 {\n",
       "  background-color: #3b4cc0;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_9758f\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_9758f_level0_col0\" class=\"col_heading level0 col0\" >Gene</th>\n",
       "      <th id=\"T_9758f_level0_col1\" class=\"col_heading level0 col1\" >Ct_Control</th>\n",
       "      <th id=\"T_9758f_level0_col2\" class=\"col_heading level0 col2\" >Ct_Treated</th>\n",
       "      <th id=\"T_9758f_level0_col3\" class=\"col_heading level0 col3\" >dCt</th>\n",
       "      <th id=\"T_9758f_level0_col4\" class=\"col_heading level0 col4\" >mean_dCt</th>\n",
       "      <th id=\"T_9758f_level0_col5\" class=\"col_heading level0 col5\" >ddCt</th>\n",
       "      <th id=\"T_9758f_level0_col6\" class=\"col_heading level0 col6\" >fold_change</th>\n",
       "      <th id=\"T_9758f_level0_col7\" class=\"col_heading level0 col7\" >High</th>\n",
       "      <th id=\"T_9758f_level0_col8\" class=\"col_heading level0 col8\" >Low</th>\n",
       "      <th id=\"T_9758f_level0_col9\" class=\"col_heading level0 col9\" >regulation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_9758f_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_9758f_row0_col0\" class=\"data row0 col0\" >GRN</td>\n",
       "      <td id=\"T_9758f_row0_col1\" class=\"data row0 col1\" >27.500000</td>\n",
       "      <td id=\"T_9758f_row0_col2\" class=\"data row0 col2\" >26.000000</td>\n",
       "      <td id=\"T_9758f_row0_col3\" class=\"data row0 col3\" >-1.500000</td>\n",
       "      <td id=\"T_9758f_row0_col4\" class=\"data row0 col4\" >-0.680000</td>\n",
       "      <td id=\"T_9758f_row0_col5\" class=\"data row0 col5\" >-0.820000</td>\n",
       "      <td id=\"T_9758f_row0_col6\" class=\"data row0 col6\" >1.765406</td>\n",
       "      <td id=\"T_9758f_row0_col7\" class=\"data row0 col7\" >True</td>\n",
       "      <td id=\"T_9758f_row0_col8\" class=\"data row0 col8\" >False</td>\n",
       "      <td id=\"T_9758f_row0_col9\" class=\"data row0 col9\" >up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_9758f_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_9758f_row1_col0\" class=\"data row1 col0\" >APP</td>\n",
       "      <td id=\"T_9758f_row1_col1\" class=\"data row1 col1\" >23.100000</td>\n",
       "      <td id=\"T_9758f_row1_col2\" class=\"data row1 col2\" >21.800000</td>\n",
       "      <td id=\"T_9758f_row1_col3\" class=\"data row1 col3\" >-1.300000</td>\n",
       "      <td id=\"T_9758f_row1_col4\" class=\"data row1 col4\" >-0.680000</td>\n",
       "      <td id=\"T_9758f_row1_col5\" class=\"data row1 col5\" >-0.620000</td>\n",
       "      <td id=\"T_9758f_row1_col6\" class=\"data row1 col6\" >1.536875</td>\n",
       "      <td id=\"T_9758f_row1_col7\" class=\"data row1 col7\" >True</td>\n",
       "      <td id=\"T_9758f_row1_col8\" class=\"data row1 col8\" >False</td>\n",
       "      <td id=\"T_9758f_row1_col9\" class=\"data row1 col9\" >up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_9758f_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_9758f_row2_col0\" class=\"data row2 col0\" >C9orf72</td>\n",
       "      <td id=\"T_9758f_row2_col1\" class=\"data row2 col1\" >29.100000</td>\n",
       "      <td id=\"T_9758f_row2_col2\" class=\"data row2 col2\" >27.900000</td>\n",
       "      <td id=\"T_9758f_row2_col3\" class=\"data row2 col3\" >-1.200000</td>\n",
       "      <td id=\"T_9758f_row2_col4\" class=\"data row2 col4\" >-0.680000</td>\n",
       "      <td id=\"T_9758f_row2_col5\" class=\"data row2 col5\" >-0.520000</td>\n",
       "      <td id=\"T_9758f_row2_col6\" class=\"data row2 col6\" >1.433955</td>\n",
       "      <td id=\"T_9758f_row2_col7\" class=\"data row2 col7\" >True</td>\n",
       "      <td id=\"T_9758f_row2_col8\" class=\"data row2 col8\" >False</td>\n",
       "      <td id=\"T_9758f_row2_col9\" class=\"data row2 col9\" >up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_9758f_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_9758f_row3_col0\" class=\"data row3 col0\" >MAPT</td>\n",
       "      <td id=\"T_9758f_row3_col1\" class=\"data row3 col1\" >26.400000</td>\n",
       "      <td id=\"T_9758f_row3_col2\" class=\"data row3 col2\" >25.700000</td>\n",
       "      <td id=\"T_9758f_row3_col3\" class=\"data row3 col3\" >-0.700000</td>\n",
       "      <td id=\"T_9758f_row3_col4\" class=\"data row3 col4\" >-0.680000</td>\n",
       "      <td id=\"T_9758f_row3_col5\" class=\"data row3 col5\" >-0.020000</td>\n",
       "      <td id=\"T_9758f_row3_col6\" class=\"data row3 col6\" >1.013959</td>\n",
       "      <td id=\"T_9758f_row3_col7\" class=\"data row3 col7\" >False</td>\n",
       "      <td id=\"T_9758f_row3_col8\" class=\"data row3 col8\" >False</td>\n",
       "      <td id=\"T_9758f_row3_col9\" class=\"data row3 col9\" >up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_9758f_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_9758f_row4_col0\" class=\"data row4 col0\" >SNCA</td>\n",
       "      <td id=\"T_9758f_row4_col1\" class=\"data row4 col1\" >24.900000</td>\n",
       "      <td id=\"T_9758f_row4_col2\" class=\"data row4 col2\" >26.200000</td>\n",
       "      <td id=\"T_9758f_row4_col3\" class=\"data row4 col3\" >1.300000</td>\n",
       "      <td id=\"T_9758f_row4_col4\" class=\"data row4 col4\" >-0.680000</td>\n",
       "      <td id=\"T_9758f_row4_col5\" class=\"data row4 col5\" >1.980000</td>\n",
       "      <td id=\"T_9758f_row4_col6\" class=\"data row4 col6\" >0.253490</td>\n",
       "      <td id=\"T_9758f_row4_col7\" class=\"data row4 col7\" >False</td>\n",
       "      <td id=\"T_9758f_row4_col8\" class=\"data row4 col8\" >True</td>\n",
       "      <td id=\"T_9758f_row4_col9\" class=\"data row4 col9\" >down</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x232e1639b10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Day 4 Challenge — “Mock qPCR Gene Expression Analysis”\n",
    "# Goal:\n",
    "# load a small dataset of mock qPCR results, clean it, analyze differential expression, and export a summary report using Pandas.\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "print(pd.__version__)\n",
    "\n",
    "PCRdata = {\n",
    "    \"Gene\": [\"APP\", \"MAPT\", \"SNCA\", \"GRN\", \"C9orf72\"],\n",
    "    \"Ct_Control\": [23.1, 26.4, 24.9, 27.5, 29.1],\n",
    "    \"Ct_Treated\": [21.8, 25.7, 26.2, 26.0, 27.9]\n",
    "}\n",
    "PCRdf = pd.DataFrame(PCRdata)\n",
    "print(PCRdf)\n",
    "print()\n",
    "\n",
    "#Compute ΔCt and ΔΔCt\n",
    "PCRdf[\"dCt\"] = PCRdf[\"Ct_Treated\"] - PCRdf[\"Ct_Control\"]\n",
    "PCRdf[\"mean_dCt\"] = PCRdf[\"dCt\"].mean()\n",
    "PCRdf[\"ddCt\"] = PCRdf[\"dCt\"] - PCRdf[\"mean_dCt\"]\n",
    "PCRdf[\"fold_change\"] = (np.power(2, -PCRdf[\"ddCt\"]))   # np.power(base, exponent) --> exponent calculation from numpy library\n",
    "\n",
    "print(\"4 new columns added: \")\n",
    "print(PCRdf)\n",
    "print()\n",
    "\n",
    "PCRdf[\"High\"] = PCRdf[\"fold_change\"] > 1.2\n",
    "PCRdf[\"Low\"] = PCRdf[\"fold_change\"] < 0.8\n",
    "\n",
    "\n",
    "PCRdf.style.background_gradient(\n",
    "    subset=[\"fold_change\"], cmap=\"coolwarm\"\n",
    ")\n",
    "\n",
    "print(\"Check for high or low fold change: \")\n",
    "print(PCRdf)\n",
    "print()\n",
    "\n",
    "PCRdf.sort_values(by=\"fold_change\", ascending=False, inplace=True)\n",
    "PCRdf.reset_index(drop=True, inplace=True)\n",
    "PCRdf[\"fold_change\"].round(2)\n",
    "\n",
    "PCRdf[\"regulation\"] = np.where(\n",
    "    PCRdf[\"fold_change\"] > 1, \"up\",\n",
    "    np.where(PCRdf[\"fold_change\"] < 1, \"down\", \"no change\")\n",
    ")\n",
    "\n",
    "print(\"Sorted by fold change and re-indexed: \")\n",
    "print()\n",
    "\n",
    "PCRdf.to_csv(\"E:/DATA_SCIENCE/Projects/qPCR_gene_expression.csv\", index=False)\n",
    "PCRdf_filtered = pd.read_csv(\"E:/DATA_SCIENCE/Projects/qPCR_gene_expression.csv\")\n",
    "print(f\"qPCR DataFrame recalled from saved .csv file:\")\n",
    "print(PCRdf_filtered)\n",
    "\n",
    "from IPython.display import display\n",
    "display(PCRdf.style.background_gradient(subset=[\"fold_change\"], cmap=\"coolwarm\"))\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
